File: windows/KikuKegel.Windows/KegelSession.cs
using System.Text.Json;

namespace KikuKegel.Windows;

internal enum ReminderMode
{
    Idle,
    Reminding,
    Exercising
}

internal enum KegelPhase
{
    Contract,
    Relax
}

internal sealed class KegelSession
{
    private const int ReminderIntervalMinutes = 45;
    private const int ContractSeconds = 5;
    private const int RelaxSeconds = 5;
    private const int CycleCount = 12;
    private readonly string _settingsPath = StoragePaths.File("kegel-session.json");
    private DateTime _nextExerciseTickAt = DateTime.MinValue;

    public event EventHandler? StateChanged;
    public event EventHandler? ReminderDue;
    public event EventHandler? ExerciseFinished;

    public ReminderMode Mode { get; private set; } = ReminderMode.Idle;
    public KegelPhase Phase { get; private set; } = KegelPhase.Contract;
    public int SecondsLeftInPhase { get; private set; } = ContractSeconds;
    public int CompletedCycles { get; private set; }
    public DateTime NextReminderAt { get; private set; }

    public int TotalCycles => CycleCount;
    public int TotalExerciseSeconds => CycleCount * (ContractSeconds + RelaxSeconds);

    public int RemainingExerciseSeconds
    {
        get
        {
            var cycleSeconds = ContractSeconds + RelaxSeconds;
            var remainingCurrentCycle = SecondsLeftInPhase + (Phase == KegelPhase.Contract ? RelaxSeconds : 0);
            var remainingFullCycles = Math.Max(0, CycleCount - CompletedCycles - 1) * cycleSeconds;
            return remainingCurrentCycle + remainingFullCycles;
        }
    }

    public string ShortStatusText
    {
        get
        {
            if (Mode != ReminderMode.Exercising)
            {
                return string.Empty;
            }

            return $"{(Phase == KegelPhase.Contract ? "收" : "放")} {SecondsLeftInPhase}s";
        }
    }

    public double ExerciseProgress
    {
        get
        {
            if (TotalExerciseSeconds <= 0)
            {
                return 0;
            }

            return Math.Clamp(1 - RemainingExerciseSeconds / (double)TotalExerciseSeconds, 0, 1);
        }
    }

    public void Start()
    {
        NextReminderAt = LoadNextReminderAt() ?? DateTime.Now.AddMinutes(ReminderIntervalMinutes);
        if (NextReminderAt <= DateTime.Now)
        {
            TriggerReminderNow();
        }
        else
        {
            NotifyChanged();
        }
    }

    public void Tick(DateTime now)
    {
        if (Mode == ReminderMode.Idle && now >= NextReminderAt)
        {
            TriggerReminderNow();
            return;
        }

        if (Mode != ReminderMode.Exercising || now < _nextExerciseTickAt)
        {
            return;
        }

        AdvanceExercise(now);
    }

    public void TriggerReminderNow()
    {
        Mode = ReminderMode.Reminding;
        ReminderDue?.Invoke(this, EventArgs.Empty);
        NotifyChanged();
    }

    public void StartExercise()
    {
        Mode = ReminderMode.Exercising;
        Phase = KegelPhase.Contract;
        SecondsLeftInPhase = ContractSeconds;
        CompletedCycles = 0;
        _nextExerciseTickAt = DateTime.Now.AddSeconds(1);
        ClearPersistedReminder();
        NotifyChanged();
    }

    public void Snooze(int minutes)
    {
        Mode = ReminderMode.Idle;
        ScheduleNextReminder(DateTime.Now.AddMinutes(minutes - ReminderIntervalMinutes));
        NotifyChanged();
    }

    private void AdvanceExercise(DateTime now)
    {
        _nextExerciseTickAt = now.AddSeconds(1);

        if (SecondsLeftInPhase > 1)
        {
            SecondsLeftInPhase -= 1;
            NotifyChanged();
            return;
        }

        if (Phase == KegelPhase.Contract)
        {
            Phase = KegelPhase.Relax;
            SecondsLeftInPhase = RelaxSeconds;
            NotifyChanged();
            return;
        }

        CompletedCycles += 1;
        if (CompletedCycles >= CycleCount)
        {
            StopAndReschedule();
            return;
        }

        Phase = KegelPhase.Contract;
        SecondsLeftInPhase = ContractSeconds;
        NotifyChanged();
    }

    private void StopAndReschedule()
    {
        Mode = ReminderMode.Idle;
        ScheduleNextReminder(DateTime.Now);
        ExerciseFinished?.Invoke(this, EventArgs.Empty);
        NotifyChanged();
    }

    private void ScheduleNextReminder(DateTime from)
    {
        NextReminderAt = from.AddMinutes(ReminderIntervalMinutes);
        PersistNextReminderAt();
    }

    private DateTime? LoadNextReminderAt()
    {
        try
        {
            if (!File.Exists(_settingsPath))
            {
                return null;
            }

            var state = JsonSerializer.Deserialize<SessionState>(File.ReadAllText(_settingsPath));
            return state?.NextReminderAt;
        }
        catch
        {
            return null;
        }
    }

    private void PersistNextReminderAt()
    {
        var state = new SessionState { NextReminderAt = NextReminderAt };
        File.WriteAllText(_settingsPath, JsonSerializer.Serialize(state));
    }

    private void ClearPersistedReminder()
    {
        if (File.Exists(_settingsPath))
        {
            File.Delete(_settingsPath);
        }
    }

    private void NotifyChanged() => StateChanged?.Invoke(this, EventArgs.Empty);

    private sealed class SessionState
    {
        public DateTime NextReminderAt { get; set; }
    }
}

File: Sources/KikuKegel/KegelSession.swift
import Foundation

enum ReminderMode: String {
    case idle
    case reminding
    case preparing
    case exercising
    case completing
}

enum KegelPhase: String {
    case contract = "收缩"
    case relax = "放松"
}

struct KegelTiming {
    var reminderInterval: TimeInterval = 45 * 60
    var contractSeconds: Int = 5
    var relaxSeconds: Int = 5
    var cycles: Int = 12
}

final class KegelSession: ObservableObject {
    private let nextReminderStorageKey = "KegelSession.nextReminderAt.v1"

    @Published private(set) var mode: ReminderMode = .idle
    @Published private(set) var phase: KegelPhase = .contract
    @Published private(set) var secondsLeftInPhase: Int = 5
    @Published private(set) var completedCycles: Int = 0
    @Published private(set) var nextReminderAt: Date
    @Published private(set) var transitionText: String = ""

    private(set) var timing = KegelTiming()
    var scheduleMode: ReminderScheduleMode = .weekdays

    var onReminder: (() -> Void)?
    var onTick: (() -> Void)?
    var onFinishExercise: (() -> Void)?

    private var reminderTimer: Timer?
    private var exerciseTimer: Timer?
    private var transitionTimer: Timer?
    private let preparationSteps = ["3", "2", "1", "开始"]
    private var preparationStepIndex = 0

    init() {
        nextReminderAt = UserDefaults.standard.object(forKey: nextReminderStorageKey) as? Date
            ?? Date().addingTimeInterval(timing.reminderInterval)
    }

    var totalCycles: Int {
        timing.cycles
    }

    var totalExerciseSeconds: Int {
        timing.cycles * (timing.contractSeconds + timing.relaxSeconds)
    }

    var remainingExerciseSeconds: Int {
        let cycleSeconds = timing.contractSeconds + timing.relaxSeconds
        let remainingCurrentCycle = secondsLeftInPhase + (phase == .contract ? timing.relaxSeconds : 0)
        let remainingFullCycles = max(0, timing.cycles - completedCycles - 1) * cycleSeconds
        return remainingCurrentCycle + remainingFullCycles
    }

    var shortStatusText: String {
        switch mode {
        case .idle:
            return ""
        case .reminding:
            return ""
        case .preparing, .completing:
            return transitionText
        case .exercising:
            return "\(phase == .contract ? "收" : "放") \(secondsLeftInPhase)s"
        }
    }

    var isExerciseFlowActive: Bool {
        mode == .preparing || mode == .exercising || mode == .completing
    }

    func start() {
        restoreOrScheduleNextReminder()
    }

    func applyScheduleMode(_ mode: ReminderScheduleMode) {
        scheduleMode = mode
        guard self.mode == .idle || self.mode == .reminding else { return }
        restoreOrScheduleNextReminder()
    }

    func applyReminderInterval(_ interval: KegelReminderInterval) {
        timing.reminderInterval = interval.timeInterval
        guard mode == .idle || mode == .reminding else { return }
        scheduleNextReminder(from: Date())
    }

    func triggerReminderNow() {
        guard scheduleMode.isActive(at: Date()) else {
            scheduleNextReminder(from: Date().addingTimeInterval(-timing.reminderInterval))
            return
        }

        reminderTimer?.invalidate()
        reminderTimer = nil
        mode = .reminding
        onReminder?()
        onTick?()
    }

    func startExercise() {
        reminderTimer?.invalidate()
        reminderTimer = nil
        exerciseTimer?.invalidate()
        transitionTimer?.invalidate()
        UserDefaults.standard.removeObject(forKey: nextReminderStorageKey)

        phase = .contract
        secondsLeftInPhase = timing.contractSeconds
        completedCycles = 0
        preparationStepIndex = 0
        transitionText = preparationSteps[preparationStepIndex]
        mode = .preparing
        onTick?()

        transitionTimer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] _ in
            self?.advancePreparationCue()
        }
    }

    func snooze(minutes: Int) {
        transitionTimer?.invalidate()
        transitionTimer = nil
        exerciseTimer?.invalidate()
        exerciseTimer = nil
        mode = .idle
        transitionText = ""
        scheduleNextReminder(from: Date().addingTimeInterval(TimeInterval(minutes * 60) - timing.reminderInterval))
        onTick?()
    }

    func pauseReminders(until date: Date) {
        reminderTimer?.invalidate()
        reminderTimer = nil
        transitionTimer?.invalidate()
        transitionTimer = nil
        exerciseTimer?.invalidate()
        exerciseTimer = nil
        mode = .idle
        transitionText = ""
        nextReminderAt = date
        persistNextReminderAt()
        let delay = max(1, date.timeIntervalSinceNow)
        let timer = Timer(timeInterval: delay, repeats: false) { [weak self] _ in
            self?.triggerReminderNow()
        }
        reminderTimer = timer
        RunLoop.main.add(timer, forMode: .common)
        onTick?()
    }

    func stopAndReschedule() {
        transitionTimer?.invalidate()
        transitionTimer = nil
        exerciseTimer?.invalidate()
        exerciseTimer = nil
        transitionText = "完成"
        mode = .completing
        onFinishExercise?()
        onTick?()

        transitionTimer = Timer.scheduledTimer(withTimeInterval: 1.35, repeats: false) { [weak self] _ in
            self?.finishCompletionCue()
        }
    }

    private func advancePreparationCue() {
        guard mode == .preparing else { return }
        preparationStepIndex += 1
        guard preparationStepIndex < preparationSteps.count else {
            beginExerciseCountdown()
            return
        }

        transitionText = preparationSteps[preparationStepIndex]
        onTick?()
    }

    private func beginExerciseCountdown() {
        guard mode == .preparing else { return }
        transitionTimer?.invalidate()
        transitionTimer = nil
        transitionText = ""
        mode = .exercising
        phase = .contract
        secondsLeftInPhase = timing.contractSeconds
        completedCycles = 0
        onTick?()

        exerciseTimer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] _ in
            self?.advanceExercise()
        }
    }

    private func finishCompletionCue() {
        guard mode == .completing else { return }
        transitionTimer?.invalidate()
        transitionTimer = nil
        transitionText = ""
        mode = .idle
        scheduleNextReminder(from: Date())
        onTick?()
    }

    private func scheduleNextReminder(from date: Date) {
        reminderTimer?.invalidate()
        let target = scheduleMode.nextReminderDate(after: date, interval: timing.reminderInterval)
        nextReminderAt = target
        persistNextReminderAt()
        let delay = max(1, target.timeIntervalSinceNow)

        let timer = Timer(timeInterval: delay, repeats: false) { [weak self] _ in
            self?.triggerReminderNow()
        }
        reminderTimer = timer
        RunLoop.main.add(timer, forMode: .common)
    }

    private func restoreOrScheduleNextReminder() {
        reminderTimer?.invalidate()
        if !scheduleMode.isActive(at: Date()), nextReminderAt <= Date() {
            scheduleNextReminder(from: Date().addingTimeInterval(-timing.reminderInterval))
            return
        }

        if nextReminderAt <= Date() {
            triggerReminderNow()
            return
        }

        let delay = max(1, nextReminderAt.timeIntervalSinceNow)
        persistNextReminderAt()
        let timer = Timer(timeInterval: delay, repeats: false) { [weak self] _ in
            self?.triggerReminderNow()
        }
        reminderTimer = timer
        RunLoop.main.add(timer, forMode: .common)
        onTick?()
    }

    private func persistNextReminderAt() {
        UserDefaults.standard.set(nextReminderAt, forKey: nextReminderStorageKey)
    }

    private func advanceExercise() {
        guard mode == .exercising else { return }

        if secondsLeftInPhase > 1 {
            secondsLeftInPhase -= 1
            onTick?()
            return
        }

        switch phase {
        case .contract:
            phase = .relax
            secondsLeftInPhase = timing.relaxSeconds
        case .relax:
            completedCycles += 1
            if completedCycles >= timing.cycles {
                stopAndReschedule()
                return
            }
            phase = .contract
            secondsLeftInPhase = timing.contractSeconds
        }

        onTick?()
    }
}
